File: algorithmic-toolbox/w4/binary_search/binary_search.py
# Uses python3
import sys

def binary_search(a, x):
    print('='*5,x,'='*5)
    left, right = 0, len(a)
    while left < right:
        mid = int((left+right)/2)
        print('mid', mid)
        if a[mid] < x:
            left = mid+1
        elif a[mid] > x: 
            right = mid
        else:
            return mid
    return -1

def linear_search(a, x):
    for i in range(len(a)):
        if a[i] == x:
            return i
    return -1

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n = data[0]
    m = data[n + 1]
    a = data[1 : n + 1]
    for x in data[n + 2:]:
        # replace with the call to binary_search when implemented
        print(binary_search(a, x), end = ' ')

File: algorithmic-toolbox/w4/binary_search/binary_search.cpp
#include <iostream>
#include <cassert>
#include <vector>

using std::vector;
using namespace std;

int binary_search(const vector<int> &a, int x) {
  // cout << "\n===== " << x << " =====\n"; 
  int left = 0, right = (int)a.size(); 
  int mid;
  while(left < right){
    mid = (int)((left+right)/2);
    // cout << "mid [" << mid << "] = " << a[mid] << "\n";
    if(a[mid] < x){
      left = mid+1;
      // cout << "wiekszy\n";
    }
    else if(a[mid] > x){
      right = mid;
      // cout << "mniejszy\n";
    }
    else{
      // cout << "dokladnie\n";
      return mid;
    }
  }
  return -1;
}

int linear_search(const vector<int> &a, int x) {
  for (size_t i = 0; i < a.size(); ++i) {
    if (a[i] == x) return i;
  }
  return -1;
}

int main() {
  int n;
  std::cin >> n;
  vector<int> a(n);
  for (size_t i = 0; i < a.size(); i++) {
    std::cin >> a[i];
  }
  int m;
  std::cin >> m;
  vector<int> b(m);
  for (int i = 0; i < m; ++i) {
    std::cin >> b[i];
  }
  for (int i = 0; i < m; ++i) {
    //replace with the call to binary_search when implemented
    std::cout << binary_search(a, b[i]) << ' ';
  }
}
